Automigration: Remove `@storybook/addon-essentials` proper

ndelangen opened the pull request:
Closes https://github.com/storybookjs/storybook/issues/30994
Closes https://github.com/storybookjs/storybook/issues/31004

## What I did

This pull request includes significant updates to the migration logic for handling the removal of `@storybook/addon-essentials` and its features moving to the core of Storybook. The changes ensure that the configuration and dependencies are correctly updated based on whether the user had the docs feature enabled or disabled.

Key changes include:

### Migration Logic Updates:
* Updated migration description to reflect that `@storybook/addon-essentials` is being removed and its features are moving to the core.
* Modified the prompt messages to provide clearer instructions based on whether the docs feature was enabled or disabled.
* Ensured that the `@storybook/addon-essentials` entry is completely removed from the configuration.
* Added logic to remove the `@storybook/addon-essentials` package and, if necessary, install `@storybook/addon-docs` if the docs feature was enabled and not already installed.

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->
This pull request has been released as version `0.0.0-pr-31015-sha-a186233f`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-31015-sha-a186233f sandbox` or in an existing project with `npx storybook@0.0.0-pr-31015-sha-a186233f upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-31015-sha-a186233f`](https://npmjs.com/package/storybook/v/0.0.0-pr-31015-sha-a186233f) |
| **Triggered by** | @valentinpalkovic |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`norbert/improve-cpc-automigrations`](https://github.com/storybookjs/storybook/tree/norbert/improve-cpc-automigrations) |
| **Commit** | [`a186233f`](https://github.com/storybookjs/storybook/commit/a186233f1e58de57bee8c3676c03cd86afc9ac6b) |
| **Datetime** | Wed Apr  2 18:09:24 UTC 2025 (`1743617364`) |
| **Workflow run** | [14226524759](https://github.com/storybookjs/storybook/actions/runs/14226524759) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=31015`_
</details>
<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->
<!-- BENCHMARK_SECTION -->


<!-- greptile_comment -->

## Greptile Summary

This PR enhances the automigration process by refining the removal logic for @storybook/addon-essentials and introduces a new migration for @storybook/addon-mdx-gfm. The changes improve configuration cleanup and dependency management based on the docs feature configuration.

- Updated /code/lib/cli-storybook/src/automigrate/fixes/addon-essentials-remove-docs.ts to fully remove addon-essentials and conditionally install addon-docs.
- Revised tests in /code/lib/cli-storybook/src/automigrate/fixes/addon-essentials-remove-docs.test.ts to verify removal and dependency commands.
- Added migration and tests for @storybook/addon-mdx-gfm in /code/lib/cli-storybook/src/automigrate/fixes/addon-mdx-gfm-remove.ts and its test file.
- Integrated the new mdx-gfm migration fix into /code/lib/cli-storybook/src/automigrate/fixes/index.ts.



<!-- /greptile_comment -->

ndelangen committed:
Refactor migration logic for @storybook/addon-essentials removal

Updated the migration process to handle the removal of @storybook/addon-essentials and its features, clarifying the installation of @storybook/addon-docs based on user configuration. The prompt messages were also improved for better user guidance during the migration.

ndelangen committed:
Add migration for removing @storybook/addon-mdx-gfm

Introduced a new migration to handle the removal of @storybook/addon-mdx-gfm from the configuration and package.json. This includes checks for its presence in the main configuration and appropriate removal actions during the migration process. Updated tests to ensure correct functionality of the migration logic.

ndelangen committed:
Merge branch 'next' into norbert/improve-cpc-automigrations

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/addon-mdx-gfm-remove.ts:
```diff
@@ -0,0 +1,85 @@
+  },
+
+  async run({ result, dryRun, packageManager, mainConfigPath }) {
+    const { hasMdxGfm } = result;
+
+    if (!hasMdxGfm) {
+      return;
+    }
```
In the `run` function `hasMdxGfm` is always true. I think also, that you can adjust the type in line 10 to `hasMdxGfm: true;`. Because if `mdxGfm` cannot be detected, you are returning null;

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/addon-mdx-gfm-remove.ts:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    await updateMainConfig({ mainConfigPath, dryRun: !!dryRun }, async (main) => {
+      // Get the current addons array
+      const addons = main.getFieldValue(['addons']) || [];
+
+      // Find the mdx-gfm entry
+      const mdxGfmIndex = addons.findIndex((addon: any) => {
+        if (typeof addon === 'string') {
+          return addon.includes('@storybook/addon-mdx-gfm');
+        }
+        return addon?.name.includes('@storybook/addon-mdx-gfm');
+      });
+
+      // Remove the mdx-gfm entry completely
+      if (mdxGfmIndex !== -1) {
+        main.removeField(['addons', mdxGfmIndex]);
+      }
+    });
+
+    if (!dryRun) {
+      // Remove addon-mdx-gfm package
+      await packageManager.removeDependencies({}, ['@storybook/addon-mdx-gfm']);
+    }
+  },
```
Isn't it easier to call `npx storybook remove @storybook/addon-mdx-gfm` instead?

ndelangen replied:
sure!

ndelangen committed:
Add migration to add Storybook overrides to package.json files

Introduced a new migration that adds Storybook overrides to package.json files when using npm. This migration checks for the presence of Storybook dependencies and modifies the package.json files accordingly. Added tests to validate the functionality of the migration process.

ndelangen committed:
Merge branch 'norbert/improve-cpc-automigrations' of https://github.com/storybookjs/storybook into norbert/improve-cpc-automigrations

ndelangen committed:
Merge branch 'next' into norbert/improve-cpc-automigrations

ndelangen committed:
Merge branch 'next' into norbert/improve-cpc-automigrations

ndelangen committed:
Remove addStorybookOverrides migration and related tests

This commit deletes the addStorybookOverrides migration and its associated test files. The migration was responsible for adding Storybook overrides to package.json files when using npm. The removal is part of a refactor to streamline the migration process and eliminate unnecessary complexity.

ndelangen committed:
Merge branch 'norbert/improve-cpc-automigrations' of https://github.com/storybookjs/storybook into norbert/improve-cpc-automigrations

ndelangen committed:
Merge branch 'next' into norbert/improve-cpc-automigrations

ndelangen committed:
Update Storybook override assignment in package.json to use variable reference

ndelangen committed:
Merge branch 'norbert/improve-cpc-automigrations' of https://github.com/storybookjs/storybook into norbert/improve-cpc-automigrations

ndelangen committed:
Merge branch 'next' into norbert/improve-cpc-automigrations

ndelangen committed:
Add migration for removing @storybook/addon-storysource

- Implemented `addonStorysourceRemove` migration to check for and remove the @storybook/addon-storysource from main configuration and package.json.
- Added corresponding tests to validate the migration functionality.
- Updated the index to include the new migration in the list of fixes.

valentinpalkovic committed:
Add migration for removing @storybook/addon-essentials-docs to index file

valentinpalkovic committed:
Refactor migration to remove @storybook/addon-essentials package

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/addon-mdx-gfm-remove.ts:
```diff
@@ -0,0 +1,58 @@
+    try {
+      let hasMdxGfm = false;
+
+      await updateMainConfig({ mainConfigPath, dryRun: true }, () => {
+        const addonNames = getAddonNames(mainConfig);
+        hasMdxGfm = addonNames.includes('@storybook/addon-mdx-gfm');
+      });
```
The `updateMainConfig` wrapper seems to be unnecessary

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/addon-mdx-gfm-remove.ts:
```diff
@@ -0,0 +1,58 @@
+    `;
+  },
+
+  async run({ result, packageManager }) {
```
`result` not used

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/addon-storysource-remove.ts:
```diff
@@ -0,0 +1,58 @@
+
+  prompt() {
+    return dedent`
+      We've detected that you have ${picocolors.yellow('@storybook/addon-storysource')} installed.
+      
+      This package has been removed in Storybook 9.0. We'll remove it from your configuration and dependencies.
```
You should mention the [migration entry](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#migration-from-storybookaddon-storysource-to-code-panel) for more information

ndelangen committed:
Refactor tests for addon-essentials migration to remove unnecessary mocks and streamline assertions

ndelangen committed:
Merge branch 'next' into norbert/improve-cpc-automigrations

ndelangen committed:
Update migration message for @storybook/addon-storysource removal to include link to migration guide

ndelangen committed:
Merge branch 'norbert/improve-cpc-automigrations' of https://github.com/storybookjs/storybook into norbert/improve-cpc-automigrations

ndelangen committed:
Refactor addonMdxGfmRemove migration to simplify run method by removing unused result parameter

ndelangen committed:
Refactor addonMdxGfmRemove migration to eliminate unused variable and streamline logic for checking addon presence
